File: Assignment_2/diabetes/mobile/lib/api_client.dart
// REST client for the diabetes screening API. The mobile app performs no inference
// itself — every screen is a view over these endpoints.
//
// Base URL: pass --dart-define=API_URL=http://<host>:8000 at build/run time.
// Default 10.0.2.2 is the Android emulator's alias for the host machine's localhost.

import 'dart:convert';
import 'package:http/http.dart' as http;

import 'models.dart';

class ApiException implements Exception {
  final String message;
  ApiException(this.message);
  @override
  String toString() => message;
}

class ApiClient {
  static const String baseUrl =
      String.fromEnvironment('API_URL', defaultValue: 'http://10.0.2.2:8000');

  final http.Client _http = http.Client();

  Uri _u(String path, [Map<String, dynamic>? q]) =>
      Uri.parse('$baseUrl$path').replace(
        queryParameters: q?.map((k, v) => MapEntry(k, '$v')),
      );

  Future<bool> health() async {
    try {
      final r = await _http
          .get(_u('/healthz'))
          .timeout(const Duration(seconds: 5));
      return r.statusCode == 200;
    } catch (_) {
      return false;
    }
  }

  Future<List<Question>> questions() async {
    final r = await _http.get(_u('/questions'));
    _check(r);
    final body = jsonDecode(r.body) as Map<String, dynamic>;
    return (body['questions'] as List)
        .map((q) => Question.fromJson((q as Map).cast<String, dynamic>()))
        .toList();
  }

  Future<Map<String, dynamic>> modelInfo() async {
    final r = await _http.get(_u('/model-info'));
    _check(r);
    return jsonDecode(r.body) as Map<String, dynamic>;
  }

  Future<PredictResult> predict(
    Map<String, dynamic> payload, {
    String include = 'explain,similar,whatif,counterfactual',
  }) async {
    final r = await _http.post(
      _u('/predict', {'include': include}),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(payload),
    );
    if (r.statusCode == 422) {
      throw ApiException('The server rejected the input: ${r.body}');
    }
    _check(r);
    return PredictResult.fromJson(
        jsonDecode(r.body) as Map<String, dynamic>);
  }

  Future<List<HistoryItem>> history(String session) async {
    final r = await _http.get(_u('/history', {'session': session}));
    _check(r);
    final body = jsonDecode(r.body) as Map<String, dynamic>;
    return (body['screenings'] as List)
        .map((h) => HistoryItem.fromJson((h as Map).cast<String, dynamic>()))
        .toList();
  }

  void _check(http.Response r) {
    if (r.statusCode >= 400) {
      throw ApiException('API error ${r.statusCode}: ${r.body}');
    }
  }
}

File: Assignment_2/diabetes/web/legacy_streamlit/api_client.py
"""Thin HTTP wrapper around the diabetes screening API for the Streamlit app."""
from __future__ import annotations

import os
from typing import Any, Optional

import requests

API_URL = os.environ.get("DIABETES_API_URL", "http://localhost:8000").rstrip("/")
TIMEOUT = 30


def _get(path: str, **params) -> Any:
    r = requests.get(f"{API_URL}{path}", params=params, timeout=TIMEOUT)
    r.raise_for_status()
    return r.json()


def _post(path: str, body: dict, headers: Optional[dict] = None, **params) -> Any:
    r = requests.post(f"{API_URL}{path}", json=body, params=params,
                      headers=headers or {}, timeout=TIMEOUT)
    if r.status_code == 422:
        raise ValueError(r.json())
    r.raise_for_status()
    return r.json()


def health() -> dict:
    return _get("/healthz")


def model_info() -> dict:
    return _get("/model-info")


def questions() -> dict:
    return _get("/questions")


def predict(payload: dict, include: str = "explain,similar,whatif,counterfactual") -> dict:
    return _post("/predict", payload, include=include)


def history(session: str, limit: int = 50) -> dict:
    return _get("/history", session=session, limit=limit)


def metrics() -> dict:
    return _get("/metrics")


def threshold_curve(split: str = "val", grid: int = 49) -> dict:
    return _get("/threshold-curve", split=split, grid=grid)


def threshold_recommend(objective: str = "recall", target: float = 0.90) -> dict:
    return _get("/threshold-recommend", objective=objective, target=target)


def get_config() -> dict:
    return _get("/config")


def set_thresholds(moderate_cut: float, high_cut: float, admin_key: str) -> dict:
    return _post("/config/thresholds",
                 {"moderate_cut": moderate_cut, "high_cut": high_cut},
                 headers={"X-Admin-Key": admin_key})
